src/KeyRegistry.sol

```diff
@@ -0,0 +1,325 @@
+     * @custom:param key    Bytes of the signer's public key.
+     * @custom:param signer Signer struct including state and merkle root for frozen signers.
+     */
+    mapping(uint256 fid => mapping(uint256 scope => mapping(bytes key => Signer signer))) public signers;
+
+    /*//////////////////////////////////////////////////////////////
+                               CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Set the IdRegistry, migration grace period, and owner.
+     *
+     * @param _idRegistry  IdRegistry contract address. Immutable.
+     * @param _gracePeriod Migration grace period in seconds. Immutable.
+     * @param _owner       Contract owner address.
+     */
+    constructor(address _idRegistry, uint24 _gracePeriod, address _owner) {
+        _transferOwnership(_owner);
+
+        gracePeriod = _gracePeriod;
+        idRegistry = IdRegistry(_idRegistry);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                               MODIFIERS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Revert if caller does not own the associated fid.
+     */
+    modifier onlyFidOwner(uint256 fid) {
+        if (idRegistry.idOf(msg.sender) != fid) revert Unauthorized();
+        _;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                              VIEWS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Retrieve signer state for a fid/scope/key tuple.
+     *
+     * @param fid   The fid associated with the key.
+     * @param scope The key's numeric scope. In the initial migration all keys will have scope 1.
+     * @param key   Bytes of the signer's public key.
+     *
+     * @return Signer struct including state and message merkle root for frozen signers.
+     */
+    function signerOf(uint256 fid, uint256 scope, bytes calldata key) external view returns (Signer memory) {
+        return signers[fid][scope][key];
+    }
+
+    /**
+     * @notice Check if the contract has been migrated.
+     *
+     * @return true if the contract has been migrated, false otherwise.
+     */
+    function isMigrated() public view returns (bool) {
+        return signersMigratedAt != 0;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                    EXTERNAL KEY REGISTRATION
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Register a public key to a fid/scope pair, setting the signer state to AUTHORIZED.
+     *
+     * @param fid   The fid associated with the key. Caller must own the provided fid.
+     * @param scope The key's numeric scope.
+     * @param key   Bytes of the signer's public key to authorize.
+     */
+    function register(uint256 fid, uint256 scope, bytes calldata key) external onlyFidOwner(fid) {
```

Comment:
it might be useful to have an additional field called `metadata` which is optional and emitted with the event, containing additional detail about the signer, like a signature from the application that requested it

Reply:
we probably also want a "register on behalf of" here

Reply:
tracked here: https://github.com/farcasterxyz/contracts/issues/263